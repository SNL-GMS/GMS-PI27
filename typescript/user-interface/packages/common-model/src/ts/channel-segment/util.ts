import isEqual from 'lodash/isEqual';

import type { ChannelSegmentTypes } from '../common-model';
import type { ChannelSegmentDescriptor } from './types';

/**
 * Determine if channelSegmentDescriptor is in the selected waveforms list
 *
 * @return returns is in list
 */
export function isSelectedWaveform(
  channelSegmentDescriptor: ChannelSegmentTypes.ChannelSegmentDescriptor,
  selectedWaveforms: ChannelSegmentTypes.ChannelSegmentDescriptor[]
): boolean {
  return (
    !!selectedWaveforms?.length &&
    selectedWaveforms.some(selectedCSD => isEqual(selectedCSD, channelSegmentDescriptor))
  );
}

/**
 * Create a unique string
 *
 * @param id ChannelSegmentDescriptor throws an error if undefined
 * @returns unique string representing the ChannelSegmentDescriptor
 */
export function createChannelSegmentString(id: ChannelSegmentDescriptor | undefined): string {
  if (!id) throw new Error('Cannot build channel segment id for undefined');
  return `${id.channel.name}.${id.channel.effectiveAt}.${id.creationTime}.${id.startTime}.${id.endTime}`;
}
