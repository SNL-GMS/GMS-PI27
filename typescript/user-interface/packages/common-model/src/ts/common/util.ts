import type { Nullable } from '../type-util/type-util';
import type { TimeRange } from './types';

/**
 * Assertion function that throws if value is nullish.
 *
 * @throws if value is nullish
 *
 * @param value a value to validate
 * @param errorMessage The error to throw if value is nullish
 * @param callback optional callback function to call before throwing the error. This may
 * be useful to toast, for example.
 */
export function validateNonNullish<T>(
  value: T | null | undefined,
  errorMessage: string,
  callback?: () => void
): asserts value is NonNullable<T> {
  if (value == null) {
    if (callback) callback();
    throw new Error(errorMessage);
  }
}

/**
 * Asserts that a {@link TimeRange} is NonNullable, and contains valid, numeric, non-null values
 *
 * @throws if timeRange is nullish
 * @throws if startTimeSecs or endTimeSecs are nullish
 * @throws if startTimeSecs or endTimeSecs are non numeric
 *
 * @param timeRange a TimeRange to check
 */
export function validateTimeRange(
  timeRange: Nullable<TimeRange> | null | undefined
): asserts timeRange is NonNullable<TimeRange> {
  validateNonNullish(
    timeRange,
    `Invalid TimeRange. TimeRange must be defined, but it is: ${timeRange}`
  );
  if (timeRange.startTimeSecs == null || timeRange.endTimeSecs == null) {
    throw new Error(`Invalid TimeRange: ${JSON.stringify(timeRange)}`);
  }
  if (typeof timeRange.startTimeSecs !== 'number' || typeof timeRange.endTimeSecs !== 'number') {
    throw new Error(`Invalid TimeRange. Values are not numeric: ${JSON.stringify(timeRange)}`);
  }
}
