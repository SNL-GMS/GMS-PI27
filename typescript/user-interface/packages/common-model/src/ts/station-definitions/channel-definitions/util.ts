import type { ChannelSegmentDescriptor } from '../../channel-segment';
import type { Channel } from './channel-definitions';
import {
  CHANNEL_COMPONENT_SEPARATOR,
  COMPONENT_SEPARATOR,
  SPLIT_CHANNEL_TOKEN,
  TEMPORARY_CHANNEL_CODE
} from './channel-definitions';

/**
 * Determines if a channel name is formatted like a rotated channel by checking for `/rotate/`
 * This should work even for rotated channels which have since been filtered.
 *
 * @param inputChannelName
 * @returns true if '/rotate/' is found
 */
export function isRotatedChannelName(inputChannelName: string): boolean {
  return inputChannelName.includes('/rotate/');
}

/**
 * Determines if a channel name is formatted like a derived channel by checking the name for a component separator token.
 *
 * @param inputChannelName the source channel name
 * @returns true if the inputChannel is derived
 */
export function isDerivedChannelName(inputChannelName: string): boolean {
  return inputChannelName.includes(COMPONENT_SEPARATOR);
}

/**
 * Determines if a channel is derived by checking the name for a component separator token.
 *
 * @param inputChannel the source channel
 * @returns true if the inputChannel is derived
 */
export function isDerivedChannel(inputChannel: ChannelSegmentDescriptor['channel']): boolean {
  return isDerivedChannelName(inputChannel.name);
}

/**
 * Determines if a channel name is formatted like a split channel by checking the name for a component separator token.
 *
 * @param inputChannelName the source channel name
 * @returns true if the inputChannel is split
 */
export function isSplitChannelName(inputChannelName: string): boolean {
  return inputChannelName.includes(SPLIT_CHANNEL_TOKEN);
}

/**
 * Determines if a channel is split by checking the name for a component separator token.
 *
 * @param inputChannel the source channel
 * @returns true if the inputChannel is split
 */
export function isSplitChannel(inputChannel: ChannelSegmentDescriptor['channel']): boolean {
  return isSplitChannelName(inputChannel.name);
}

/**
 * Determines if a channel name is temporary by the temp channel code
 *
 * @param inputChannelName the source channel name
 * @returns true if the inputChannelName is temporary
 */
export function isTemporaryChannelName(inputChannelName: string): boolean {
  return inputChannelName !== undefined && inputChannelName?.includes(TEMPORARY_CHANNEL_CODE);
}

/**
 * Determines if a channel is temporary by checking the name for a temporary channel code.
 *
 * @param inputChannel the source channel
 * @returns true if the inputChannel is temporary
 */
export function isTemporaryChannel(inputChannel: Channel): boolean {
  return isTemporaryChannelName(inputChannel?.name);
}

/**
 * Determines if a channel name is formatted like a raw channel by checking the name for a component separator token.
 *
 * @param inputChannelName the source channel name
 * @returns true if the inputChannel is raw
 */
export function isRawChannelName(inputChannelName: string | undefined): boolean {
  return (
    !!inputChannelName &&
    ((!inputChannelName.includes('/') && inputChannelName.includes(CHANNEL_COMPONENT_SEPARATOR)) ||
      (inputChannelName.includes(CHANNEL_COMPONENT_SEPARATOR) &&
        !inputChannelName.includes('beam') &&
        !isRotatedChannelName(inputChannelName)))
  );
}

/**
 * Determines if a channel is raw by checking the name for a component separator token.
 *
 * @param inputChannel the source channel
 * @returns true if the inputChannel is raw
 */
export function isRawChannel(
  inputChannel: ChannelSegmentDescriptor['channel'] | undefined
): boolean {
  return !!inputChannel && isRawChannelName(inputChannel.name);
}

/**
 * Returns the beam type or undefined parsed from channel name
 *
 * @param channelName
 * @returns string beam type
 */
export function parseWaveformChannelType(
  channelName: string | undefined
): 'N/A' | 'Raw channel' | 'Fk beam' | 'Event beam' | 'Detection beam' | undefined {
  if (!channelName) {
    return undefined;
  }

  if (isTemporaryChannelName(channelName)) {
    return 'N/A';
  }

  if (isRawChannelName(channelName)) {
    return 'Raw channel';
  }

  const elements = channelName.split('/');

  if (!elements[1]) {
    return undefined;
  }

  const beamDescription = elements[1] === 'masked' ? elements[2] : elements[1];

  const beamPrefix = beamDescription.split(',')[1];
  switch (beamPrefix) {
    case 'fk':
      return 'Fk beam';
    case 'event':
      return 'Event beam';
    case 'detection':
      return 'Detection beam';
    default:
      return undefined;
  }
}
