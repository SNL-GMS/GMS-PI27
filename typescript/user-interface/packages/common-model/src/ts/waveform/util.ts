import type { Timeseries } from '../channel-segment';
import type { WaveformTypes } from '../common-model';

/**
 * TypeGuard to check type of a Timeseries to see if it is a {@link Waveform}.
 *
 * @param maybeWaveform a Timeseries to check if it is a Waveform timeseries and assert the type
 */
export function isWaveformTimeseries(
  maybeWaveform: Timeseries
): maybeWaveform is WaveformTypes.Waveform {
  return (
    (maybeWaveform as WaveformTypes.Waveform).samples != null ||
    (maybeWaveform as WaveformTypes.Waveform)._uiClaimCheckId != null
  );
}
