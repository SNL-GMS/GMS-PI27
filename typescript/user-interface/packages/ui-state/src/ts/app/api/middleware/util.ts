import { UILogger } from '@gms/ui-util';
import type { SerializedAxiosError } from '@gms/ui-workers';
import type { Action } from '@reduxjs/toolkit';
import type {
  ListenerErrorHandler,
  ListenerErrorInfo
} from '@reduxjs/toolkit/dist/listenerMiddleware/types';
import includes from 'lodash/includes';

import { selectOpenIntervalName } from '../../state/workflow/selectors';
import type { AppState } from '../../store';

/** returns the environment variable string that is used for data population logging and debugging */
export const GMS_DATA_POPULATION = 'GMS_DATA_POPULATION' as const;

/** returns environment variable setting for the data population logging and debugging  */
export const ENV_GMS_DATA_POPULATION = process.env[GMS_DATA_POPULATION];

export const dataPopulationLogger = UILogger.create(GMS_DATA_POPULATION, ENV_GMS_DATA_POPULATION);

const INFO = 'Data Population Middleware:' as const;

export const dataPopulationOnError: ListenerErrorHandler = (
  error: unknown,
  errorInfo: ListenerErrorInfo
) => {
  dataPopulationLogger.error(`${INFO} error occurred`, error, errorInfo);
};

export function isIntervalOpen(state: AppState) {
  const openIntervalName = selectOpenIntervalName(state);

  return openIntervalName != null && openIntervalName !== '';
}

/**
 * @example isRejectedAction(action,
 *                           [`${getEventsWithDetectionsAndSegmentsByTime.typePrefix}/rejected`])
 *
 * @param action to be checked for failure
 * @param rejectedActions array of all rejected actions that the caller wants to identify
 * @returns true if {action} is included in {rejectedActions} array
 */
export function isRejectedAction<T = unknown>(
  action: Action,
  rejectedActions: string[]
): action is {
  type: string;
  payload: SerializedAxiosError;
  error: { message: string };
  meta: { arg: T };
} {
  return includes(rejectedActions, action.type);
}
