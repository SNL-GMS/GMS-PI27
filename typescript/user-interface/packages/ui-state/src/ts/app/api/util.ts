import { batch } from 'react-redux';

import type { AppDispatch } from '../store';
import { dataSlice } from './data';
import { eventManagerApiSlice } from './event-manager/event-manager-api-slice';
import { processingConfigurationApiSlice } from './processing-configuration';
import { userManagerApiSlice } from './user-manager';
import { workflowApiSlice } from './workflow';

/**
 * Clears out the entire api state.
 *
 * @param dispatch
 */
export const resetApiState = (dispatch: AppDispatch): void => {
  batch(() => {
    dispatch(eventManagerApiSlice.util.resetApiState());
    dispatch(workflowApiSlice.util.resetApiState());
    dispatch(userManagerApiSlice.util.resetApiState());
    dispatch(processingConfigurationApiSlice.util.resetApiState());
    dispatch(dataSlice.actions.clearAll());
  });
};
