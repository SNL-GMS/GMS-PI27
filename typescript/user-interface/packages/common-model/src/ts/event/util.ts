import { getDistance, getGreatCircleBearing } from 'geolib';
import memoize from 'lodash/memoize';

import { findOrThrow } from '../array-util/array-util';
import type { CommonTypes, EventTypes, SignalDetectionTypes } from '../common-model';
import { getCurrentHypothesis } from '../signal-detection/util';
import type { Event, EventHypothesis, EventLocation, LocationSolution } from './types';

/**
 * Finds preferred event hypothesis for a given event and stage
 * If no preferred hypothesis exists for stage, will return most recent preferred hypothesis by stage
 *
 * @param event Event to search within
 * @param openIntervalName Currently open interval/stage name eg; "AL1"
 * @returns EventHypothesis corresponding to the current open stage, otherwise default stage EventHypothesis.
 */
// TODO: This should NOT return undefined; requires refactoring for other components
export const findPreferredEventHypothesisByOpenStageOrDefaultStage = (
  event: Event | undefined,
  openIntervalName: string
): EventHypothesis | undefined => {
  if (!event) {
    return undefined;
  }
  const { preferredEventHypothesisByStage, eventHypotheses } = event;

  const preferredHypoByStageId =
    preferredEventHypothesisByStage.find(hypo => hypo.stage.name === openIntervalName)?.preferred ??
    preferredEventHypothesisByStage[0].preferred; // Default preferred

  // Should not be capable of returning undefined
  // Should either return hypo by the open stage, or hypo by auto network (default)
  return findOrThrow<EventTypes.EventHypothesis>(
    eventHypotheses,
    hypothesis => hypothesis.id.hypothesisId === preferredHypoByStageId.id.hypothesisId
  );
};

/**
 * Determines if a Signal Detection is associated to an event by searching event hypothesis for
 * the SD id. Stops scanning when the first association is found.
 *
 * @param currentSDHypothesis the current hypothesis from the SD to be checked for event association
 * @param events the events you want to search through to check for SD association
 * @param openIntervalName
 * @returns
 */
export function isSignalDetectionUnassociated(
  sdHypo: SignalDetectionTypes.SignalDetectionHypothesis,
  events: EventTypes.Event[],
  openIntervalName: string
) {
  return !events.find(event => {
    const preferredEventHypothesis = findPreferredEventHypothesisByOpenStageOrDefaultStage(
      event,
      openIntervalName
    );
    if (preferredEventHypothesis) {
      return preferredEventHypothesis.associatedSignalDetectionHypotheses.find(
        associatedSignalDetectionHypothesis => associatedSignalDetectionHypothesis.id === sdHypo.id
      );
    }
    return false;
  });
}

/**
 * Determine if a signal detection is associated to an event.
 *
 * @param detection signal detection to check event association against
 * @param events list of {@link EventTypes.Event} objects to search through
 * @param currentOpenEventId
 * @param openIntervalName eg; Auto Network, AL1, AL2
 * @returns boolean
 */
export function isSignalDetectionOpenAssociated(
  detection: SignalDetectionTypes.SignalDetection,
  events: EventTypes.Event[],
  currentOpenEventId: string,
  openIntervalName: string
): boolean {
  if (!currentOpenEventId || !events) {
    return false;
  }
  const currentOpenEvent = events.find(event => event.id === currentOpenEventId);
  const currentPreferredHypothesis = findPreferredEventHypothesisByOpenStageOrDefaultStage(
    currentOpenEvent,
    openIntervalName
  );

  // Determines if the current SD hypothesis is in the associated SD
  // list for the current preferred event hypothesis
  return !!currentPreferredHypothesis?.associatedSignalDetectionHypotheses?.find(
    sd => sd?.id.id === getCurrentHypothesis(detection?.signalDetectionHypotheses)?.id.id
  );
}

/**
 * Determines if the provided {@link EventHypothesis} object is the
 * preferredEventHypothesis for the current open stage.
 *
 * @param event Event object to check
 * @param openIntervalName Current open stage
 * @param eventHypothesis EventHypothesis object to verify
 * @returns true if the eventHypothesis is preferred
 */
export const isPreferredEventHypothesisByStage = (
  event: Event,
  openIntervalName: string,
  eventHypothesis: EventHypothesis
): boolean => {
  if (!event || !eventHypothesis) {
    return false;
  }

  // Get the preferredEventHypo exists for the open stage (if it exists)
  const maybePreferred = event.preferredEventHypothesisByStage.find(
    hypo => hypo.stage.name === openIntervalName
  );

  // Check if these two have the same identifiers
  return (
    maybePreferred?.preferred.id.eventId === eventHypothesis.id.eventId &&
    maybePreferred?.preferred.id.hypothesisId === eventHypothesis.id.hypothesisId
  );
};

/**
 * Finds the last non rejected parent event hypothesis for a given event and hypothesis
 * if no valid hypothesis is found it returns the first hypothesis
 *
 * @param event
 * @param eventHypothesis
 */
export const findEventHypothesisParent = (
  event: Event,
  /** Can be either the hypothesis for the open stage, or the default stage */
  eventHypothesis: EventHypothesis | undefined
): EventHypothesis | undefined => {
  if (!eventHypothesis) {
    return undefined;
  }

  // loop backwards until we find a non-rejected hypothesis
  for (let i = eventHypothesis.parentEventHypotheses.length - 1; i >= 0; i -= 1) {
    const parentEventHypothesis = event.eventHypotheses.find(
      hypo => hypo.id.hypothesisId === eventHypothesis.parentEventHypotheses[i].id.hypothesisId
    );
    if (parentEventHypothesis) {
      return parentEventHypothesis;
    }
  }

  // nothing was found, return the first hypothesis
  return event.eventHypotheses[0];
};

/**
 * Finds the preferred location solution for a hypothesis falling back to the parents if the hypothesis is rejected
 *
 * @param eventHypothesisId hypothesis id to find the solution for
 * @param eventHypotheses list of hypotheses, if an event is opened
 * @returns a location solution
 */
export const findPreferredLocationSolution = (
  eventHypothesisId: string,
  eventHypotheses: EventHypothesis[]
): LocationSolution | undefined => {
  const eventHypothesis = eventHypotheses.find(
    hypothesis => hypothesis.id.hypothesisId === eventHypothesisId
  );
  if (!eventHypothesis) return undefined;
  if (eventHypothesis.preferredLocationSolution) {
    return eventHypothesis.locationSolutions.find(
      ls => ls.id === eventHypothesis?.preferredLocationSolution?.id
    );
  }

  if (eventHypothesis.parentEventHypotheses) {
    const parentHypothesisId =
      eventHypothesis.parentEventHypotheses[eventHypothesis.parentEventHypotheses.length - 1].id
        .hypothesisId;

    const parentEventHypothesis = eventHypotheses.find(
      hypothesis => hypothesis.id.hypothesisId === parentHypothesisId
    );

    return parentEventHypothesis?.locationSolutions.find(
      ls => ls.id === parentEventHypothesis?.preferredLocationSolution?.id
    );
  }

  return undefined;
};

/**
 * Calculate the distance in kilometers between an event and a station.
 *
 * @param source The source location for which to calculate distance
 * @param destination The destination location for which to calculate distance
 * @returns calculated distance in kilometers
 */
export function getLocationToEventDistance(
  source: CommonTypes.Location,
  destination: EventLocation
): CommonTypes.Distance {
  const accuracy = 1000;
  const degreePrecision = 1000;
  const KM = 1000;

  const KM_TO_DEGREES = 111.1949266;
  const dist: number = getDistance(
    { latitude: source.latitudeDegrees, longitude: source.longitudeDegrees },
    {
      latitude: destination.latitudeDegrees,
      longitude: destination.longitudeDegrees
    },
    accuracy
  );
  const kmDistance = dist / KM;
  // return distance as degrees and km
  return {
    degrees: Math.round((kmDistance / KM_TO_DEGREES) * degreePrecision) / degreePrecision,
    km: kmDistance
  };
}

/**
 * Calculate a location to event location azimuth
 *
 * @param source The source location for which to calculate azimuth
 * @param destination The destination location for which to calculate azimuth
 * @returns calculated distance in kilometers
 */
export function getLocationToLocationAzimuth(
  source: CommonTypes.Location,
  destination: EventLocation
): number {
  const origin = {
    latitude: source.latitudeDegrees,
    longitude: source.longitudeDegrees
  };
  const dest = {
    latitude: destination.latitudeDegrees,
    longitude: destination.longitudeDegrees
  };
  return getGreatCircleBearing(origin, dest);
}

// A function to explicitly set the memoization cache ID.
// This resolves an issue where the cache hitting a false positive
const memoizeRecordResolver = (source, destination) =>
  `${source.latitudeDegrees},${source.longitudeDegrees}; ${destination.latitudeDegrees},${destination.longitudeDegrees}`;

export const memoizedLocationToEventDistance = memoize(
  getLocationToEventDistance,
  memoizeRecordResolver
);

export const memoizedLocationToEventAzimuth = memoize(
  getLocationToLocationAzimuth,
  memoizeRecordResolver
);
