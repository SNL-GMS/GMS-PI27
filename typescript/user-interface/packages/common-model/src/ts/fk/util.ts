import type { ChannelSegment, Timeseries } from '../channel-segment/types';
import { TimeseriesType } from '../channel-segment/types';
import type { FkTypes, SignalDetectionTypes } from '../common-model';
import { findPhaseFeatureMeasurementValue, getCurrentHypothesis } from '../signal-detection/util';
import type { FkFrequencyRange, FkPowerSpectra, FkSpectra } from './types';

/** Kilometer per degree of Earth */
const KM_PER_DEGREE = 111.19492664;

/**
 * Checks if FK spectra channel segment
 *
 * @param object Channel Segment
 * @returns boolean
 */
export function isFkSpectraChannelSegment(
  object: ChannelSegment<Timeseries>
): object is ChannelSegment<FkSpectra> {
  return object.timeseriesType === TimeseriesType.FK_SPECTRA;
}

/**
 * Checks if FK spectra
 *
 * @param object FkSpectra Timeseries
 * @returns boolean
 */
export function isFkSpectra(object: Timeseries): object is FkSpectra {
  return object.type === TimeseriesType.FK_SPECTRA;
}

/**
 * Checks if FK power spectra channel segment
 *
 * @param object Channel Segment
 * @returns boolean
 */
export function isFkPowerSpectraChannelSegment(
  object: ChannelSegment<Timeseries>
): object is ChannelSegment<FkSpectra> {
  return object.timeseriesType === TimeseriesType.FK_SPECTRA_OLD;
}

/**
 * Checks if FK spectra
 *
 * @param object FkSpectra Timeseries
 * @returns boolean
 */
export function isFkPowerSpectra(object: Timeseries): object is FkPowerSpectra {
  return object.type === TimeseriesType.FK_SPECTRA_OLD;
}

// TODO: Review to see what might make more sense as part of configuration
/** Width/height of y and x axis respectively */
export const SIZE_OF_FK_RENDERING_AXIS_PX = 35;

/** The height of the rendering needs to be 8 pixels smaller than the width */
export const FK_RENDERING_HEIGHT_OFFSET = 8;
/** The width of the x axis markers needs to be 10 pixels wider than the canvas */
export const FK_X_AXIS_WIDTH_OFFSET = 10;
/** The height of the y axis markers needs to be 12 pixels taller than the canvas */
export const FK_Y_AXIS_HEIGHT_OFFSET = 12;

export const BEAM_TOOLBAR_WIDTH_OFFSET = -80;

/** Default Frequency Bands */
export const FrequencyBands: FkFrequencyRange[] = [
  {
    lowFrequencyHz: 0.5,
    highFrequencyHz: 2
  },
  {
    lowFrequencyHz: 1,
    highFrequencyHz: 2.5
  },
  {
    lowFrequencyHz: 1.5,
    highFrequencyHz: 3
  },
  {
    lowFrequencyHz: 2,
    highFrequencyHz: 4
  },
  {
    lowFrequencyHz: 3,
    highFrequencyHz: 6
  }
];

/**
 * Formats a frequency band into a string for the drop down
 *
 * @param band Frequency band to format
 */
export function frequencyBandToString(band: FkTypes.FkFrequencyRange): string {
  return `${band.lowFrequencyHz} - ${band.highFrequencyHz} Hz`;
}

/**
 * Creates menu options for frequency bands
 */
export const generateFrequencyBandOptions = (): string[] => {
  const items: string[] = [];
  FrequencyBands.forEach(frequency => {
    items.push(frequencyBandToString(frequency));
  });
  return items;
};

/**
 * Approximate conversion between degrees and km
 *
 * @returns Kilometer
 */
export function degreeToKmApproximate(degree: number): number {
  const DEGREES_IN_CIRCLE = 360;
  const RAD_EARTH = 6371;
  const TWO_PI = Math.PI * 2;
  return degree / (DEGREES_IN_CIRCLE / (RAD_EARTH * TWO_PI));
}

/**
 * Approximate conversion between km and degrees
 *
 * @returns Degree
 */
export function kmToDegreesApproximate(km: number): number {
  const DEGREES_IN_CIRCLE = 360;
  const RAD_EARTH = 6371;
  const TWO_PI = Math.PI * 2;
  return km * (DEGREES_IN_CIRCLE / (RAD_EARTH * TWO_PI));
}

/**
 * @returns Input value in s/degree converted to km/s
 */
export function convertSecondsPerDegreeToKilometersPerSecond(value: number) {
  return KM_PER_DEGREE / value;
}

/**
 * @returns Input value in s/degree converted to s/km
 */
export function convertSecondsPerDegreeToSecondsPerKilometer(value: number) {
  return 1 / convertSecondsPerDegreeToKilometersPerSecond(value);
}

/**
 * Determines if the fk needs review
 *
 * @param sd: the signal detection to check
 * @returns true if needs review, otherwise false
 */
export const fkNeedsReview = (
  fk: FkTypes.FkSpectra | undefined,
  phasesNeedingReview: string[] | undefined,
  sd: SignalDetectionTypes.SignalDetection | undefined,
  isForFiltering = false
): boolean => {
  if (!fk || !sd || !phasesNeedingReview) {
    return false;
  }
  if (fk.reviewed && !isForFiltering) {
    return false;
  }
  const currentSdHypothesis = getCurrentHypothesis(sd.signalDetectionHypotheses);
  const phase = findPhaseFeatureMeasurementValue(currentSdHypothesis.featureMeasurements)?.value;

  if (phase && phasesNeedingReview.indexOf(phase) > -1) {
    return true;
  }
  return false;
};
