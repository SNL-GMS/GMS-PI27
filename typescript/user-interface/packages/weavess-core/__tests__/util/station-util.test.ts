import type { Channel, Station } from '../../src/ts/types';
import { findChannelInStation, findChannelInStations } from '../../src/ts/util';
import { stations } from './__data__/weavess-stations';

describe('test station util', () => {
  test('can find parent and child channels', () => {
    // Test ability to find Weavess Channel based on channel name
    const station: Station | undefined = stations[0];
    const channel: Channel | undefined = station?.nonDefaultChannels
      ? station.nonDefaultChannels[0]
      : undefined;
    expect(station).toBeDefined();
    expect(channel).toBeDefined();

    expect(findChannelInStation(station, channel?.id ?? '')).toMatchSnapshot();
    expect(findChannelInStation(station, station.id)).toMatchSnapshot();
  });

  it('should have function findWeavessChannel', () => {
    // Test finding a station
    let channel: Channel | undefined = findChannelInStations(stations, 'PDAR');
    expect(channel).toBeDefined();
    expect(channel?.id).toEqual('PDAR');

    // Test finding a channel
    channel = findChannelInStations(stations, 'PDAR.PD10.SHZ');
    expect(channel).toBeDefined();
    expect(channel?.id).toEqual('PDAR.PD10.SHZ');

    // Test finding a bogus channel
    channel = findChannelInStations(stations, 'FOO');
    expect(channel).toBeUndefined();
  });
});
