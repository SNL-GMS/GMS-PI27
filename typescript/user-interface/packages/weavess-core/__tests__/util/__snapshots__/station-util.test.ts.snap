// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test station util can find parent and child channels 1`] = `
{
  "height": 24.8,
  "id": "AKASG.AKASG.HHZ",
  "isSelected": false,
  "name": "name",
  "timeOffsetSeconds": 0,
  "waveform": {
    "channelSegmentId": "unfiltered",
    "channelSegmentsRecord": {},
    "markers": {
      "verticalMarkers": [
        {
          "color": "#fff",
          "id": "startTime",
          "lineStyle": "solid",
          "timeSecs": 1274392801,
        },
        {
          "color": "#fff",
          "id": "endTime",
          "lineStyle": "solid",
          "timeSecs": 1274400000,
        },
      ],
    },
    "masks": [],
    "signalDetections": [],
  },
}
`;

exports[`test station util can find parent and child channels 2`] = `
{
  "height": 24.8,
  "id": "AKASG",
  "isSelected": false,
  "name": "name",
  "timeOffsetSeconds": 0,
  "waveform": {
    "channelSegmentId": "unfiltered",
    "channelSegmentsRecord": {},
    "markers": {
      "selectionWindows": [],
      "verticalMarkers": [
        {
          "color": "#fff",
          "id": "startTime",
          "lineStyle": "solid",
          "timeSecs": 1274392801,
        },
        {
          "color": "#fff",
          "id": "endTime",
          "lineStyle": "solid",
          "timeSecs": 1274400000,
        },
      ],
    },
    "predictedPhases": [],
    "signalDetections": [],
  },
}
`;
